src/bsoncxx/document/element.hpp

```diff
@@ -31,73 +31,104 @@ enum class type : std::uint8_t;
-        types::b_null get_null() const;
-        types::b_regex get_regex() const;
-        types::b_dbpointer get_dbpointer() const;
-        types::b_code get_code() const;
-        types::b_symbol get_symbol() const;
-        types::b_codewscope get_codewscope() const;
-        types::b_int32 get_int32() const;
-        types::b_timestamp get_timestamp() const;
-        types::b_int64 get_int64() const;
-        types::b_minkey get_minkey() const;
-        types::b_maxkey get_maxkey() const;
-
-        types::value get_value() const;
-
-        const std::uint8_t* raw;
-        std::uint32_t length;
-        std::uint32_t offset;
-    };
+class BSONCXX_API element {
+   public:
+    element();
+
+    explicit element(const std::uint8_t* raw, std::uint32_t length, std::uint32_t offset);
+
+    explicit operator bool() const;
+
+    bsoncxx::type type() const;
+
+    stdx::string_view key() const;
+
+    // TODO figure out a better exception than std::runtime_error
+    types::b_double get_double() const;
+    types::b_utf8 get_utf8() const;
+    types::b_document get_document() const;
+    types::b_array get_array() const;
+    types::b_binary get_binary() const;
+    types::b_undefined get_undefined() const;
+    types::b_oid get_oid() const;
+    types::b_bool get_bool() const;
+    types::b_date get_date() const;
+    types::b_null get_null() const;
+    types::b_regex get_regex() const;
+    types::b_dbpointer get_dbpointer() const;
+    types::b_code get_code() const;
+    types::b_symbol get_symbol() const;
+    types::b_codewscope get_codewscope() const;
+    types::b_int32 get_int32() const;
+    types::b_timestamp get_timestamp() const;
+    types::b_int64 get_int64() const;
+    types::b_minkey get_minkey() const;
+    types::b_maxkey get_maxkey() const;
+
+    types::value get_value() const;
+
+    ///
+    /// If this element is a document, finds the first element of the document with the provided
+    /// key. If there is no such element, an invalid document::element will be returned. The
+    /// runtime of operator[] is linear in the length of the document.
+    ///
+    /// @throws std::runtime_error if this element is not a document.
+    ///
+    /// @param key
+    ///   The key to search for.
+    ///
+    /// @return The matching element, if found, or an invalid element.
+    ///
+    element operator[](stdx::string_view key) const;
+
+    ///
+    /// If this element is an array, indexes into this BSON array. If the index is out-of-bounds,
+    /// an invalid array::element will be returned. As BSON represents arrays as documents, the
+    /// runtime of operator[] is linear in the length of the array.
+    ///
+    /// @throws std::runtime_error if this element is not an array.
+    ///
+    /// @param i
+    ///   The index of the element.
+    ///
+    /// @return The element if it exists, or an invalid element.
+    ///
+    array::element operator[](std::uint32_t i) const;
+
+    const std::uint8_t* raw;
```

Comment:
I would like accessors better I think, but lets not change them here, as I'd like to get this pushed. Could you or @Machyne please file a ticket for -rc0 to make them private and add accessors. It seems like it would be straightforward.